impeller/aiks/paint_pass_delegate.cc

```diff
@@ -44,4 +48,88 @@ std::shared_ptr<Contents> PaintPassDelegate::CreateContentsForSubpassTarget(
                                             effect_transform);
 }
 
+/// OpacityPeepholePassDelegate
+/// ----------------------------------------------
+
+OpacityPeepholePassDelegate::OpacityPeepholePassDelegate(
+    Paint paint,
+    std::optional<Rect> coverage)
+    : paint_(std::move(paint)), coverage_(coverage) {}
+
+// |EntityPassDelgate|
+OpacityPeepholePassDelegate::~OpacityPeepholePassDelegate() = default;
+
+// |EntityPassDelgate|
+std::optional<Rect> OpacityPeepholePassDelegate::GetCoverageRect() {
+  return coverage_;
+}
+
+// |EntityPassDelgate|
+bool OpacityPeepholePassDelegate::CanElide() {
+  return paint_.blend_mode == BlendMode::kDestination;
+}
+
+// |EntityPassDelgate|
+bool OpacityPeepholePassDelegate::CanCollapseIntoParentPass(
+    EntityPass* entity_pass) {
+  // Note: determing whether any coverage intersects has quadradic complexity in
+  // the number of rectangles, and depending on whether or not we cache at
+  // different levels of the entity tree may end up cubic. In the interest of
+  // proving whether or not this optimization is valuable, we only consider very
+  // simple peephole optimizations here - where there is a single drawing
+  // command wrapped in save layer. This would indicate something like an
+  // Opacity or FadeTransition wrapping a very simple widget, like in the
+  // CupertinoPicker.
+  if (entity_pass->GetEntityCount() > 3) {
+    // Single paint command with a save layer would be:
+    // 1. clip
+    // 2. draw command
+    // 3. restore.
+    return false;
+  }
+  bool all_can_accept = true;
+  std::vector<Rect> all_coverages;
+  auto had_subpass = entity_pass->IterateAllFlatEntities([&](Entity& entity) {
+    auto contents = entity.GetContents();
+    if (!contents->CanAcceptOpacity(entity)) {
+      all_can_accept = false;
+      return false;
+    }
+    auto maybe_coverage = contents->GetCoverage(entity);
+    if (maybe_coverage.has_value()) {
+      auto coverage = maybe_coverage.value();
+      for (const auto& cv : all_coverages) {
```

Comment:
Can we make this even simpler by just bailing out if we don't get the structure in the comments above (i.e. clip, draw, restore)?

A small change above could lead to a very expensive change down here without realizing it.

Reply:
We don't have a good way to check what kind of entity something is.